### DISCUSSION TITLE
Workflow `Application` is marked as `OutOfSync/Missing` when all workflow instances are removed due to ttl

### DISCUSSION THREAD

**Discussion by masterginger** [argoproj/argo-workflows#11436] [Q&A] [answered]
We deploy Argo Workflow as an Argo Application in our k8s deployments.

When we set `ttlStrategy` in workflow's spec to 2 days, the workflow instance will be removed 2 days after it has finished.
We observed that if all existing workflow runs are removed due to ttl expiration, the Argo Workflow application itself is marked as OutOfSync/Missing.

Is there a way to disable this behavior and keep the Argo Workflow application itself in "Healthy" state?

**Comment by jeremy-fisher-nd** [comment, marked as answer]
Looks like a duplicate of https://github.com/argoproj/argo-cd/discussions/8247

**Reply by masterginger** [reply]
Re-opening this since the hook deletion policy suggested in https://github.com/argoproj/argo-cd/discussions/8247 didn't work. I tried to add the hook annotations to the workflow spec, but the workflow was deleted right after it's synced but the underlying steps were still running. As a result, the workflow execution was always interrupted because the steps in the workflow didn't have enough time to finish.

**Comment by masterginger** [comment]
@agilgur5 could you help re-open this discussion? Thanks :)

**Comment by Joibel** [comment]
I am unsure what this has to do with Argo Workflows.
ArgoCD's job is to make sure objects in kubernetes match those in git. You're somehow expecting an object in kubernetes to disappear and ArgoCD not to get upset about that.
Argo Workflows is behaving as intended here, and I don't see any change being made to workflows to alleviate your problem.

**Reply by masterginger** [reply]
Thank for your reply @Joibel ! We set the ttlStrategy to limit the number of historical workflow stored to save resources. The challenge is if there hasn't been a sync for "> ttl" time, the last workflow instance will be deleted due to ttl expiration and the corresponding Argo Application will be marked as out of sync.

Do you know if there's a way to configure the workflow as "keep the last N workflows" instead of "delete workflows from more than N seconds ago"?

**Reply by Joibel** [reply]
If you don't put a TTL on then it will follow `retentionPolicy` in the workflow controller configmap https://github.com/argoproj/argo-workflows/blob/feaed7ac8cf8a0667fc20b8be5bd590bef4da42a/docs/workflow-controller-configmap.yaml#L354.

Your problem, which is a common one, might be that you think that just because this workflow is called `foobar-abc12` and there is another workflow called `foobar-def34` that the controller may consider them both to be `foobar` workflow and be able to control retention on the set of workflows called `foobar`. Actually the controller doesn't have a concept of a set of workflows, and cannot do this.

**Reply by Joibel** [reply]
I would argue that deploying `workflows` through Argo CD is always wrong. I'm happy that deploying `workflow templates` that way is correct, but attempting to overload CD into doing that job is incorrect, and the source of your problem.

Have you considered using the same hook that's causing CD to update the application, I'll pretend this is a commit from github, and using Argo Events to deliver that to Argo Workflows to run the workflow instead? Workflows can wait for the ArgoCD application to be in the appropriate state by observing the application object in the cluster.

**Reply by masterginger** [reply]
Yes, we are deploying `workflow templates` via an Argo Application sync-ing on a Git tag in our repository (we do plan to move to Argo Events).

Thanks for the pointer to the `retentionPolicy` field, that's exactly what we were looking for!

**Reply by masterginger** [reply]
Actually `retentionPolicy` might have the same downside if there are multiple workflows running. Because `retentionPolicy` is a global setting, which means it determines the number of workflow objects to keep in the same app group, across all the namespaces. So if one of the namespaces has a much faster deployment pace than the other repos using argo workflow, the "recent workflows" will be all from that repo and the workflows from other repos will be deleted.

I guess Argo Events would be the ultimate solution as it decouples Argo Workflow from Argo App.
